### ISSUE TITLE
Fallback confuse behaviour

### ISSUE THREAD

**Issue by schumannc** [talkdai/dialog#117]
**Current Behavior:**
In the current implementation of `fallback`, when a document retrieval process yields no results, the system adds fallback text directly into the prompt body. This approach delegates the responsibility of handling empty or no-result scenarios to the LLM, which may not always align with the expected or desired output for end-users.

**Issue**
It is unclear if the current behavior was intentionally designed this way. From an outside perspective, integrating fallback text into the prompt seems to obscure the model's capacity to appropriately handle cases where no documents are retrieved.

**Suggested Improvement:**
It may be more logical and user-friendly to have a fixed fallback text or a predefined response strategy for scenarios where the document retrieval process returns no results.

**Comment by lgabs** [comment]
Here is an example asking "quem foi o presidente do Brasil em 2001?" using the `.toml` from README (assuming the db is already up) and its answer:

logs:
```
  dialog git:(main) ✗ docker compose up dialog
[+] Building 0.0s (0/0)                                                                                  docker:desktop-linux
[+] Running 2/2
 ✔ Container dialog-db-1      Recreated                                                                                  0.4s
 ✔ Container dialog-dialog-1  Recreated                                                                                  0.1s
Attaching to dialog-dialog-1
dialog-dialog-1  | INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
dialog-dialog-1  | INFO  [alembic.runtime.migration] Will assume transactional DDL.
dialog-dialog-1  | /app/src/load_csv.py:4: DeprecationWarning:
dialog-dialog-1  | Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
dialog-dialog-1  | (to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
dialog-dialog-1  | but was not found to be installed on your system.
dialog-dialog-1  | If this would cause problems for you,
dialog-dialog-1  | please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
dialog-dialog-1  |
dialog-dialog-1  |   import pandas as pd
dialog-dialog-1  | Generating embeddings for new questions...
dialog-dialog-1  | New questions: 0
dialog-dialog-1  | INFO:     Will watch for changes in these directories: ['/app/src']
dialog-dialog-1  | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
dialog-dialog-1  | INFO:     Started reloader process [44] using WatchFiles
dialog-dialog-1  | INFO:     Started server process [46]
dialog-dialog-1  | INFO:     Waiting for application startup.
dialog-dialog-1  | INFO:     Application startup complete.
dialog-dialog-1  | 2024-02-27 01:44:33,168 - INFO - HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"
dialog-dialog-1  | ================================ System Message ================================
dialog-dialog-1  |
dialog-dialog-1  | I'm sorry, I didn't understand your question. Could you rephrase it?
dialog-dialog-1  |
dialog-dialog-1  | ================================ Human Message =================================
dialog-dialog-1  |
dialog-dialog-1  | {user_message}
dialog-dialog-1  | 2024-02-27 01:44:33,220 - INFO - Verbose LLM prompt: None
dialog-dialog-1  | /usr/local/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:117: LangChainDeprecationWarning: The function `__call__` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use invoke instead.
dialog-dialog-1  |   warn_deprecated(
dialog-dialog-1  | 2024-02-27 01:44:35,417 - INFO - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
dialog-dialog-1  | 2024-02-27 01:44:35,462 - INFO - Request processing time for chat_id cb2c6ada8d6948d794144051ece695ee: 0:00:03.478990
dialog-dialog-1  | INFO:     192.168.65.1:32340 - "POST /chat/cb2c6ada8d6948d794144051ece695ee HTTP/1.1" 200 OK
dialog-dialog-1  | 2024-02-27 01:44:35,467 - INFO - keywords: cancelo, reserva, cancelar, siga, passos, abaixo, acesse, aba, viagens, selecione, deseja, final, página, seção, ajuda, clique, informe, motivo, cancelamento, conclua, processo, caso, alguma, dúvida, precise, adicional, disposição, ajudar, presidente, brasil, momento, jair, bolsonaro, 2001, fernando, henrique, cardoso, 1995, 200
```
answer:
```
{
  "message": "Em 2001, o presidente do Brasil era Fernando Henrique Cardoso. Ele foi presidente do Brasil de 1995 a 2003."
}
```

This actually _uses the LLM outside the scope of the documents_, since, after discovering there are no documents retrieved, we still build a strange prompt with the fallback phrase and send its to GPT to answer. However, I've already tested that if the prompt can follow instructions to answer some fallback phrase in case of question outside document's context or if there are none in dedicated "Context" section.

I think the solution here can go in two (at minimum) ways:

1. Avoid making a prompt to GPT and, in case of zero documents retrieved, answer the fixed fallback as @schumannc suggested
2. Prompt Engineering: delegate the decision about fallback to GPT. In this case, even with relevant documents (it can happen, it's difficult to say if a query will hit the cosine distance threshold), if GPT sees that the question is outside context, it answers accordingly. This is not a bad option either, I've tested prompts with this kind of instruction and it works (langchain's prompt hub for [rag prompt includes this instruction](https://smith.langchain.com/hub/rlm/rag-prompt))

**Comment by lgabs** [comment]
There is a second problem we have to deal/discuss with follow-up questions: when using memory history, option (1) I've said above can not work so well, for example:

first user question
```
user: how can I cancel my order?
AI: you can cancel your have to come to the store and bring the product or I can cancel it online for you and send someone to get the product [...]
```
second user question (with previous memory):
```
user: can you do this for me?
[no documents retrieved from vector db]
AI: I'm sorry, I could not understand your question. Can you rephrase it?
```
This would happen since the last question has no meaning without previous history context, but with option (2) GPT would have context to decide how to answer.

**Comment by lgabs** [comment]
Probably the default behavior should be the fixed fallback, and if the developer wants to delegate to GPT the answer using instructions from the prompt template, than he/she can make its own custom subclass of `AbstractLLM` dealing with desired behavior.
